test/unit/org/apache/cassandra/dht/BootStrapperTest.java

```diff
@@ -83,6 +106,57 @@ public void testSourceTargetComputation() throws UnknownHostException
         }
     }
 
+    @Test
+    @BMRules(rules = { @BMRule(name = "Make sure the non-optimized path is picked up for some operations",
+                               targetClass = "org.apache.cassandra.dht.RangeStreamer",
+                               targetMethod = "convertPreferredEndpointsToWorkMap(EndpointsByReplica)",
+                               action = "org.apache.cassandra.dht.BootStrapperTest.nonOptimizationHit()"),
+                       @BMRule(name = "Make sure the optimized path is picked up for some operations",
+                               targetClass = "org.apache.cassandra.dht.RangeStreamer",
+                               targetMethod = "getOptimizedWorkMap(EndpointsByReplica,Collection,String)",
+                               action = "org.apache.cassandra.dht.BootStrapperTest.optimizationHit()") })
+    public void testOptmizationsSkip() throws UnknownHostException
+    {
+        testSkipFFOptmizationFeatureFlag(true, true, false);
+        testSkipFFOptmizationFeatureFlag(false, true, true);
+    }
+
+    private void testSkipFFOptmizationFeatureFlag(boolean disableOptimization, boolean nonOptimizedPathHit, boolean optimizedPathHit) throws UnknownHostException
+    {
+        nonOptimizationHit.set(false);
+        optimizationHit.set(false);
+        CassandraRelevantProperties.SKIP_OPTIMAL_STREAMING_CANDIDATES_CALCULATION.setBoolean(disableOptimization);
+
+        for (String keyspaceName : Schema.instance.getNonLocalStrategyKeyspaces().names())
+        {
+            StorageService ss = StorageService.instance;
+            TokenMetadata tmd = ss.getTokenMetadata();
+
+            generateFakeEndpoints(10);
+            Token myToken = tmd.partitioner.getRandomToken();
+            InetAddressAndPort myEndpoint = InetAddressAndPort.getByName("127.0.0.1");
+
+            assertEquals(10, tmd.sortedTokens().size());
+            RangeStreamer s = new RangeStreamer(tmd, null, myEndpoint, StreamOperation.BOOTSTRAP, true, DatabaseDescriptor.getEndpointSnitch(), new StreamStateStore(), mockFailureDetector, false, 1);
+            s.addRanges(keyspaceName, Keyspace.open(keyspaceName).getReplicationStrategy().getPendingAddressRanges(tmd, myToken, myEndpoint));
+        }
+
+        assertEquals(nonOptimizedPathHit, nonOptimizationHit.get());
+        assertEquals(optimizedPathHit, optimizationHit.get());
+        CassandraRelevantProperties.SKIP_OPTIMAL_STREAMING_CANDIDATES_CALCULATION.reset();
```

Comment:
I think resetting should be in the `finally` block to make sure it gets executed even if the earlier assertions fail.

Reply:
Will do